File: day5.c
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

bool order[100][100];

int main()
{
    FILE* file = fopen("day5.txt", "r");
    int sum[2] = {};
    char line[256];

    while (fgets(line, sizeof(line), file)) {
        int x, y;
        if (2 != sscanf(line, "%d|%d", &x, &y)) break;
        order[x][y] = true;
    }

    while (fgets(line, sizeof(line), file)) {
        int pages[30], n = 0;
        char* t = strtok(line, ",\n");
        while (t) {
            sscanf(t, "%d", &pages[n++]);
            t = strtok(NULL, ",\n");
        }

        bool sorted = true;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (order[pages[j]][pages[i]]) {
                    int tmp = pages[i];
                    pages[i] = pages[j];
                    pages[j] = tmp;
                    sorted = false;
                }
            }
        }
        sum[!sorted] += pages[n / 2];
    }

    printf("%d %d\n", sum[0], sum[1]);
}

File: day5.lua
#!/usr/bin/env lua

local file = io.open("day5.txt")
local sum = { 0, 0 }
local order = {}
setmetatable(order, { __index = function(t, k) t[k] = {} return t[k] end })

for line in file:lines() do
    local x, y = line:match("(%d+)|(%d+)")
    if not x then break end
    order[tonumber(x)][tonumber(y)] = true
end

for line in file:lines() do
    local pages = {}
    for x in line:gmatch("%d+") do
        table.insert(pages, tonumber(x))
    end

    local sorted = true
    for i = 1, #pages do
        for j = i + 1, #pages do
            if order[pages[j]][pages[i]] then
                pages[i], pages[j] = pages[j], pages[i]
                sorted = false
            end
        end
    end

    local idx = sorted and 1 or 2
    sum[idx] = sum[idx] + pages[#pages // 2 + 1]
end

print(sum[1], sum[2])
